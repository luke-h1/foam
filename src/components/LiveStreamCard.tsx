import useThemeColor from '@app/hooks/useThemeColor';
import twitchService, { Stream } from '@app/services/twitchService';
import theme from '@app/styles/theme';
import { Image, ImageStyle } from 'expo-image';
import { useState, useEffect } from 'react';
import { StyleSheet, View, ViewStyle } from 'react-native';
import { TouchableOpacity } from 'react-native-gesture-handler';
import LiveStreamImage from './LiveStreamImage';
import Tags from './Tags';
import Text from './Text';

interface Props {
  stream: Stream;
}

export default function LiveStreamCard({ stream }: Props) {
  const shadow = useThemeColor({ light: theme.dropShadow, dark: undefined });
  const [broadcasterImage, setBroadcasterImage] = useState<string>();

  const getUserProfilePictures = async () => {
    const res = await twitchService.getUserImage(stream.user_login);
    setBroadcasterImage(res);
  };

  useEffect(() => {
    getUserProfilePictures();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [stream]);

  return (
    <TouchableOpacity activeOpacity={0.8}>
      <View style={[styles.streamCard, shadow]}>
        <View style={styles.streamHeadline}>
          <LiveStreamImage
            animated
            thumbnail={stream.thumbnail_url}
            startedAt={stream.started_at}
            size="large"
          />
          <View style={styles.streamDetail}>
            <Text size="xxs" weight="light">
              {stream.title}
            </Text>
            <View style={styles.streamMetadata}>
              <View style={styles.userInfo}>
                <Image
                  source={{ uri: broadcasterImage }}
                  style={styles.avatar}
                  testID="LiveStreamCard-avatar"
                />
                <Text size="xs" weight="bold">
                  {stream.user_name}
                </Text>
              </View>
              <Text size="xxs" weight="light">
                {new Intl.NumberFormat('en-US').format(stream.viewer_count)}{' '}
                viewers
              </Text>
            </View>
          </View>
        </View>

        <View style={{ marginTop: theme.spacing.xs }}>
          <Tags tags={stream.tags} />
        </View>
      </View>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create<{
  streamCard: ViewStyle;
  streamHeadline: ViewStyle;
  streamDetail: ViewStyle;
  streamMetadata: ViewStyle;
  userInfo: ViewStyle;
  avatar: ImageStyle;
}>({
  streamCard: {
    flex: 1,
    padding: theme.spacing.xs,
    marginHorizontal: theme.spacing.sm,
    borderRadius: theme.borderradii.sm,
    marginBottom: theme.spacing.md,
  },
  streamDetail: {
    flex: 1,
    justifyContent: 'flex-start',
  },
  streamHeadline: {
    flex: 1,
    flexDirection: 'row',
  },
  streamMetadata: {
    marginTop: theme.spacing.md,
    flex: 1,
    flexDirection: 'column',
  },
  userInfo: {
    flexDirection: 'row',
    alignItems: 'center',
    marginTop: 5,
  },
  avatar: {
    width: 20,
    height: 20,
    borderRadius: 13,
    marginRight: 5,
  },
});
