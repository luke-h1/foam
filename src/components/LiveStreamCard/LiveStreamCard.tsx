import { useStreamerImage } from '@app/hooks';
import { Stream } from '@app/services';
import { elapsedStreamTime, formatViewCount } from '@app/utils';
import { useRouter } from 'expo-router';
import { View } from 'react-native';
import { StyleSheet } from 'react-native-unistyles';
import { Button } from '../Button';
import { Image } from '../Image';
import { Typography } from '../Typography';

interface Props {
  stream: Stream;
}

export function LiveStreamCard({ stream }: Props) {
  const { profilePicture } = useStreamerImage(stream.user_login, [stream]);

  const router = useRouter();

  return (
    <Button
      onPress={() => {
        router.push(`/streams/${stream.id}`);
      }}
    >
      <View style={styles.container}>
        <View style={styles.imageContainer}>
          <Image
            source={stream?.thumbnail_url
              .replace('{width}', '1920')
              .replace('{height}', '1080')}
            style={{
              width: 150,
              height: 100,
            }}
            transition={150}
          />
          <View style={styles.overlay}>
            <View style={styles.redDot} />
            <Typography size="xxs">
              {elapsedStreamTime(stream.started_at)}
            </Typography>
          </View>
        </View>
        <View style={styles.details}>
          <Typography size="sm">{stream.title}</Typography>
          <View style={styles.metadata}>
            <View style={styles.info}>
              <Image
                source={profilePicture}
                style={styles.avatar}
                testID="LiveStreamCard-avatar"
              />

              <Typography size="xs">{stream.user_name}</Typography>
            </View>
            <Typography size="xxs">
              {formatViewCount(stream.viewer_count)} viewers
            </Typography>
          </View>
          <Typography size="sm">{stream.game_name}</Typography>
        </View>
      </View>
    </Button>
  );
}

const styles = StyleSheet.create(theme => ({
  container: {
    flexDirection: 'row',
    paddingVertical: theme.spacing.xl,
    paddingHorizontal: theme.spacing.sm,
    marginBottom: theme.spacing.md,
    columnGap: theme.spacing.sm,
    flex: 1,
  },
  tagWrapper: {
    width: '85%',
  },
  imageContainer: {
    position: 'relative',
  },
  overlay: {
    position: 'absolute',
    top: 3,
    left: 3,
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
    paddingHorizontal: 8,
    paddingVertical: 4,
    borderRadius: 5,
  },
  redDot: {
    width: 8,
    height: 8,
    borderRadius: 4,
    backgroundColor: 'red',
    marginRight: 5,
  },
  liveText: {
    // color: 'white',
  },
  details: {
    flex: 1,
    justifyContent: 'flex-start',
    marginLeft: theme.spacing.md,
  },
  title: {
    fontWeight: 'bold',
  },
  metadata: {
    marginVertical: theme.spacing.sm,
  },
  info: {
    flexDirection: 'row',
    alignItems: 'center',
    marginVertical: theme.spacing.md,
  },
  avatar: {
    width: 20,
    height: 20,
    borderRadius: 10,
    marginRight: theme.spacing.xs,
  },
  tagsContainer: {
    marginTop: theme.spacing.xs,
  },
}));
