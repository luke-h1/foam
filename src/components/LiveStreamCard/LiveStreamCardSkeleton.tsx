import { View } from 'react-native';
import { StyleSheet } from 'react-native-unistyles';
import { Skeleton } from '../Skeleton';

export function LiveStreamCardSkeleton() {
  return (
    <View style={styles.container}>
      <View style={styles.imageContainer}>
        <Skeleton style={styles.imageSkeleton} />
      </View>
      <View style={styles.details}>
        <Skeleton style={styles.titleSkeleton} />
        <View style={styles.metadata}>
          <View style={styles.info}>
            <Skeleton style={styles.avatarSkeleton} />
            <Skeleton style={styles.textSkeleton} />
          </View>
          <Skeleton style={styles.textSkeleton} />
        </View>
        <Skeleton style={styles.textSkeleton} />
      </View>
    </View>
  );
}

const styles = StyleSheet.create(theme => ({
  container: {
    flexDirection: 'row',
    paddingVertical: theme.spacing.xl,
    paddingHorizontal: theme.spacing.sm,
    marginBottom: theme.spacing.md,
    columnGap: theme.spacing.sm,
    flex: 1,
  },
  imageContainer: {
    position: 'relative',
  },
  imageSkeleton: {
    width: 150,
    height: 100,
    borderRadius: theme.radii.md,
  },
  details: {
    flex: 1,
    justifyContent: 'flex-start',
    marginLeft: theme.spacing.md,
  },
  titleSkeleton: {
    height: 20,
    width: '80%',
    marginBottom: theme.spacing.sm,
  },
  metadata: {
    marginVertical: theme.spacing.sm,
  },
  info: {
    flexDirection: 'row',
    alignItems: 'center',
    marginVertical: theme.spacing.md,
  },
  avatarSkeleton: {
    width: 20,
    height: 20,
    borderRadius: 10,
    marginRight: theme.spacing.xs,
    borderCurve: 'continuous',
  },
  textSkeleton: {
    height: 15,
    width: '60%',
    marginBottom: theme.spacing.xs,
  },
}));
