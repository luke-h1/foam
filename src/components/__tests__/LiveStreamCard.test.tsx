import _twitchService, { Stream } from '@app/services/twitchService';
import render from '@app/test/render';
import { screen } from '@testing-library/react-native';
import LiveStreamCard from '../LiveStreamCard';

const mockStream: Stream = {
  id: '1',
  user_id: '123',
  user_login: 'test_user',
  user_name: 'Test user',
  game_id: '456',
  game_name: 'Just Chatting',
  type: 'live',
  title: 'Test Stream Title',
  viewer_count: 10000,
  started_at: new Date().toISOString(),
  language: 'en',
  thumbnail_url: 'https://example.com/thumbnail.jpg',
  tags: ['stream', 'live', '2024', 'foam'],
  tag_ids: [],
  is_mature: false,
};

jest.mock('@app/services/twitchService');

const twitchService = jest.mocked(_twitchService);

describe('LiveStreamCard', () => {
  beforeEach(() => {
    twitchService.getUserImage.mockResolvedValue(
      'https://example.com/avatar.jpg',
    );
  });

  test('renders correctly', () => {
    render(<LiveStreamCard stream={mockStream} />);

    expect(screen.getByText('Test Stream Title')).toBeOnTheScreen();
    expect(screen.getByText('Test user')).toBeOnTheScreen();
    expect(screen.getByTestId('LiveStreamCard-avatar')).toBeOnTheScreen();
    expect(screen.getByText('10,000 viewers')).toBeOnTheScreen();
    expect(screen).toMatchSnapshot();
  });
});
