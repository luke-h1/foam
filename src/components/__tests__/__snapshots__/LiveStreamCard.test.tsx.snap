// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveStreamCard renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      {
        "columnGap": 4,
        "flex": 1,
        "flexDirection": "row",
        "marginBottom": 8,
        "paddingHorizontal": 4,
        "paddingVertical": 16,
      }
    }
  >
    <View
      style={
        {
          "position": "relative",
        }
      }
    >
      <ViewManagerAdapter_ExpoImage
        containerViewRef={"[React.ref]"}
        contentFit="cover"
        contentPosition={
          {
            "left": "50%",
            "top": "50%",
          }
        }
        height={100}
        nativeViewRef={"[React.ref]"}
        onError={[Function]}
        onLoad={[Function]}
        onLoadStart={[Function]}
        onProgress={[Function]}
        placeholder={[]}
        source={
          [
            {
              "uri": "https://example.com/thumbnail.jpg",
            },
          ]
        }
        style={
          {
            "height": 100,
            "width": 150,
          }
        }
        transition={null}
        width={150}
      />
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "rgba(0, 0, 0, 0.5)",
            "borderRadius": 5,
            "flexDirection": "row",
            "left": 3,
            "paddingHorizontal": 8,
            "paddingVertical": 4,
            "position": "absolute",
            "top": 3,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "red",
              "borderRadius": 4,
              "height": 8,
              "marginRight": 5,
              "width": 8,
            }
          }
        />
        <Text
          style={
            [
              {
                "color": "rgba(255, 255, 255, 0.8)",
                "fontFamily": "Inter",
                "fontSize": 16,
                "fontWeight": 400,
              },
              {
                "color": "white",
                "fontSize": 12,
                "fontWeight": "bold",
              },
            ]
          }
        >
          00m
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flex": 1,
          "justifyContent": "flex-start",
          "marginLeft": 8,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "rgba(255, 255, 255, 0.8)",
              "fontFamily": "Inter",
              "fontSize": 14,
              "fontWeight": 600,
            },
            undefined,
          ]
        }
      >
        Test Stream Title
      </Text>
      <View
        style={
          {
            "marginVertical": 4,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "marginVertical": 8,
            }
          }
        >
          <ViewManagerAdapter_ExpoImage
            borderRadius={10}
            containerViewRef={"[React.ref]"}
            contentFit="cover"
            contentPosition={
              {
                "left": "50%",
                "top": "50%",
              }
            }
            height={20}
            marginRight={2}
            nativeViewRef={"[React.ref]"}
            onError={[Function]}
            onLoad={[Function]}
            onLoadStart={[Function]}
            onProgress={[Function]}
            placeholder={[]}
            source={
              [
                {
                  "uri": "https://example.com/avatar.jpg",
                },
              ]
            }
            style={
              {
                "borderRadius": 10,
                "height": 20,
                "marginRight": 2,
                "width": 20,
              }
            }
            testID="LiveStreamCard-avatar"
            transition={null}
            width={20}
          />
          <Text
            style={
              [
                {
                  "color": "rgba(255, 255, 255, 0.8)",
                  "fontFamily": "Inter",
                  "fontSize": 16,
                  "fontWeight": 400,
                },
                undefined,
              ]
            }
          >
             
            Test user
          </Text>
        </View>
        <Text
          style={
            [
              {
                "color": "rgba(255, 255, 255, 0.8)",
                "fontFamily": "Inter",
                "fontSize": 12,
                "fontWeight": 400,
              },
              undefined,
            ]
          }
        >
          10,000
           viewers
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "rgba(255, 255, 255, 0.8)",
              "fontFamily": "Inter",
              "fontSize": 12,
              "fontWeight": 400,
            },
            undefined,
          ]
        }
      >
        Just Chatting
      </Text>
    </View>
  </View>
</View>
`;
